runtimeconfig/manager.go

```diff
@@ -0,0 +1,210 @@
+package runtimeconfig
+
+import (
+	"bytes"
+	"context"
+	"crypto/sha256"
+	"flag"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"sync"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+
+	"github.com/grafana/dskit/services"
+)
+
+// Loader loads the configuration from file.
+type Loader func(r io.Reader) (interface{}, error)
+
+// Config holds the config for an Manager instance.
+// It holds config related to loading per-tenant config.
+type Config struct {
+	ReloadPeriod time.Duration `yaml:"period"`
+	// LoadPath contains the path to the runtime config file, requires an
+	// non-empty value
+	LoadPath string `yaml:"file"`
+	Loader   Loader `yaml:"-"`
+}
+
+// RegisterFlags registers flags.
+func (mc *Config) RegisterFlags(f *flag.FlagSet) {
+	f.StringVar(&mc.LoadPath, "runtime-config.file", "", "File with the configuration that can be updated in runtime.")
+	f.DurationVar(&mc.ReloadPeriod, "runtime-config.reload-period", 10*time.Second, "How often to check runtime config file.")
+}
+
+// Manager periodically reloads the configuration from a file, and keeps this
+// configuration available for clients.
+type Manager struct {
+	services.Service
+
+	cfg    Config
+	logger log.Logger
+
+	listenersMtx sync.Mutex
+	listeners    []chan interface{}
+
+	configMtx sync.RWMutex
+	config    interface{}
+
+	configLoadSuccess prometheus.Gauge
+	configHash        *prometheus.GaugeVec
+}
+
+// New creates an instance of Manager and starts reload config loop based on config
+func New(cfg Config, registerer prometheus.Registerer, logger log.Logger) (*Manager, error) {
```

Comment:
Should we add a prefix here for the metric names?

Reply:
Generic is fine. People can pass custom registrerer that adds prefix, if they want to. (`prometheus.WrapRegistererWithPrefix`)